cli/config/credentials/oauth_store.go

```diff
@@ -0,0 +1,151 @@
+package credentials
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/docker/cli/cli/config/types"
+	"github.com/docker/cli/cli/oauth"
+	"github.com/docker/docker/registry"
+)
+
+// oauthStore wraps an existing store that transparently handles oauth
+// flows, managing authentication/token refresh and piggybacking off an
+// existing store for storage/retrieval.
+type oauthStore struct {
+	backingStore Store
+	manager      oauth.Manager
+}
+
+// NewOAuthStore creates a new oauthStore backed by the provided store.
+func NewOAuthStore(backingStore Store, manager oauth.Manager) Store {
+	return &oauthStore{
+		backingStore: backingStore,
+		manager:      manager,
+	}
+}
+
+const minimumTokenLifetime = 50 * time.Minute
+
+// Get retrieves the credentials from the backing store, refreshing the
+// access token if the retrieved token is valid for less than 50 minutes.
+// If there are no credentials in the backing store, the device code flow
+// is initiated with the tenant in order to log the user in and get
+func (o *oauthStore) Get(serverAddress string) (types.AuthConfig, error) {
+	if serverAddress != registry.IndexServer {
+		return o.backingStore.Get(serverAddress)
+	}
+
+	auth, err := o.backingStore.Get(serverAddress)
+	if err != nil {
+		// If an error happens here, it's not due to the backing store not
+		// containing credentials, but rather an actual issue with the backing
+		// store itself. This should be propagated up.
+		return types.AuthConfig{}, err
+	}
+
+	tokenRes, err := o.parseToken(auth.Password)
+	// if the password is not a token, return the auth config as is
+	if err != nil {
+		//nolint:nilerr
+		return auth, nil
+	}
+
+	// if the access token is valid for less than 50 minutes, refresh it
+	if tokenRes.RefreshToken != "" && tokenRes.Claims.Expiry.Time().Before(time.Now().Add(minimumTokenLifetime)) {
+		refreshRes, err := o.manager.RefreshToken(context.TODO(), tokenRes.RefreshToken)
+		if err != nil {
+			return types.AuthConfig{}, err
+		}
+		tokenRes = refreshRes
+	}
+
+	err = o.storeInBackingStore(tokenRes)
+	if err != nil {
+		return types.AuthConfig{}, err
+	}
+
+	return types.AuthConfig{
+		Username:      tokenRes.Claims.Domain.Username,
+		Password:      tokenRes.AccessToken,
+		Email:         tokenRes.Claims.Domain.Email,
+		ServerAddress: registry.IndexServer,
+	}, nil
+}
+
+// GetAll returns a map containing solely the auth config for the official
+// registry, parsed from the backing store and refreshed if necessary.
+func (o *oauthStore) GetAll() (map[string]types.AuthConfig, error) {
+	allAuths, err := o.backingStore.GetAll()
+	if err != nil {
+		return nil, err
+	}
+
+	if _, ok := allAuths[registry.IndexServer]; !ok {
+		return allAuths, nil
+	}
+
+	auth, err := o.Get(registry.IndexServer)
+	if err != nil {
+		return nil, err
+	}
+	allAuths[registry.IndexServer] = auth
```

Comment:
We want to get all the credentials, so we fetch all credentials from the backing store. If there are no configured credentials for the official registry, we can return them as is. Otherwise, we call `Get` to get that specific one, so that we can make sure the retrieved credentials are refreshed/all the same logic applies as a normal `Get`.

I'll try to document it a bit more.

It's also usually good practice to look at the tests to see if they provide you clarity on what the requirements for a piece of code are.

Reply:
Thank you!

>It's also usually good practice to look at the tests to see if they provide you clarity on what the requirements for a piece of code are.

Yes I agree, although in the context of security we should be mindful to keep code as clear as possible.